ldms/src/sampler/zfs_topvdevs/zfs_topvdevs.c

```diff
@@ -0,0 +1,440 @@
+	if (nvlist_lookup_uint64_array(	nvroot,
+					ZPOOL_CONFIG_VDEV_STATS,
+					(uint64_t **)&vs, &c) != 0) {
+                rc = 1;
+        }
+
+        record_instance = ldms_record_alloc(sampler_base->set,
+                                            index_store.vdev_rec_idx);
+
+        if (record_instance == NULL) {
+                log_fn(LDMSD_LDEBUG, SAMP": ldms_record_alloc() failed, resizing metric set\n");
+                resize_metric_set();
+		record_instance = ldms_record_alloc(sampler_base->set,
+				index_store.vdev_rec_idx);
+		if (record_instance == NULL)
+		  rc = 2;
+        }
+
+	if (rc == 0) {
+		rc = ldms_list_append_record(sampler_base->set, list_handle,
+						record_instance);
+
+		/* zpoolname    0 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[0],
+				pool_name);
+	        /* zpool state  1 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[1],
+				zpool_state_to_name((vdev_state_t)vs->vs_state, (vdev_aux_t)vs->vs_aux));
+	        /* vdevname     2 */
+		ldms_record_array_set_str(record_instance, vdev_metric_ids[2],
+						vdev_name);
+	        /* alloc        3 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[3], vs->vs_alloc);
+	        /* free         4 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[4], vs->vs_space - vs->vs_alloc);
+		/* size         5 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[5], vs->vs_space);
+		/* read_bytes   6 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[6], vs->vs_bytes[ZIO_TYPE_READ]);
+		/* iread_errors 7 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[7], vs->vs_read_errors);
+		/* read_ops     8 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[8], vs->vs_ops[ZIO_TYPE_READ]);
+		/* write_bytes  9 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[9], vs->vs_bytes[ZIO_TYPE_WRITE]);
+		/* write_errors 10 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[10], vs->vs_write_errors);
+		/* write_ops    11 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[11], vs->vs_ops[ZIO_TYPE_WRITE]);
+		/* checksum errors 12 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[12], vs->vs_checksum_errors);
+		/* fragmentation 13 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[13], vs->vs_fragmentation);
+		/* initialization errors 14 */
+		ldms_record_set_u64(record_instance, vdev_metric_ids[14], vs->vs_initialize_errors);
+	}
+
+	free(vdev_name);
+	vdev_name = NULL;
+
+        return (rc);
+}
+
+
+/*
+ * recursive stats printer
+ */
+static int get_recursive_stats(stat_printer_f func, nvlist_t *nvroot,
+    const char *pool_name, int descend, zpool_handle_t *zhp)
+{
+        uint_t c, children;
+        nvlist_t **child;
+        int err = 0;
+
+        if (descend && nvlist_lookup_nvlist_array(nvroot, ZPOOL_CONFIG_CHILDREN,
+            &child, &children) == 0) {
+                for (c = 0; c < children; c++) {
+			err = func(child[c], pool_name, zhp);
```

Comment:
This will stomp on previous errors and only return the last status (which might be success).  We should probably either 1) `break` the loop of the first error, 2) preserve the last error, or even 3) stop pretending we care about the error and always return success.  I think I'd vote for 2) given the options, but I'm not entirely sure what the sampler behavior is when an error is returned.  Does it just drop the entire sample?